src/python/pants/engine/goal.py

```diff
@@ -91,17 +115,18 @@ class List(Goal):
 
     See {doc_url('plugin-upgrade-guide')}.
     """
-    environment_migrated: ClassVar[bool] = False
+    environment_behavior: ClassVar[EnvironmentBehavior] = EnvironmentBehavior.UNMIGRATED
 
     @classmethod
     def _get_environment_migrated(cls) -> bool:
         deprecated_conditional(
-            lambda: not cls.environment_migrated,
+            lambda: cls.environment_behavior == Goal.EnvironmentBehavior.UNMIGRATED,
             "2.17.0.dev0",
-            f"Not setting `Goal.environment_migrated=True` for `Goal` `{cls.name}`",
+            f"Setting `Goal.environment_behavior=EnvironmentBehavior.UNMIGRATED` for `Goal` "
+            f"`{cls.name}`",
             hint=f"See {doc_url('plugin-upgrade-guide')}\n",
         )
-        return cls.environment_migrated
+        return cls.environment_behavior != Goal.EnvironmentBehavior.UNMIGRATED
```

Comment:
I think you want to return True if it's UNMIGRATED or LOCAL_ONLY. This method is used inside `engine_initializer.py` to know whether to inject the local environment or not, which we should do in both cases.

It probably makes sense to rename this method to something like `_needs_local_environment_injected_into_rule_graph`

Reply:
My understanding, when I looked at the upgrade guide, was that `LOCAL_ONLY` should explicitly use `ChosenLocalEnvironmentName` to find the environment name, and then pass that down through the dispatch `Get` later on. This matches the rules that have been migrated so far.

Happy to defer to you/@stuhood if my reckoning is incorrect here.

Reply:
Ah okay. So is this enum meant to be deprecated eventually? It may be worth clarifying there is no runtime difference between LOCAL_ONLY vs USES_ENVIRONMENTS, it's only meant as an FYI to the reader.

Also, I think the advice to set to LOCAL_ONLY to silence the deprecation is insufficient; you also need to set up ChosenLocalEnvironmentName.

--

Although, that all feels like unnecessary boilerplate IMO.. I think it could be sensible to have this enum set to LOCAL_ONLY to mean engine_initializer.py injects it for you, and keep that around indefinitely

Reply:
Yeah, I'm in-principle happy with actually assigning behaviour to `LOCAL_ONLY` as a path forward. I'm going to find some time to chat with @stuhood about what the path forward on the remaining unmigrated goals looks like. I suspect there will need to be an onus on goal authors to verify that they are not performing invalid operations

Reply:
I was thinking that we would continue to inject an `EnvironmentName` for local-only goals. `ChosenLocalEnvironmentName` is what is computed by `determine_bootstrap_environment`, and so marking your goal as "local only" would be a convenience to avoid making any changes to your goal.

Reply:
Cool, I think that makes sense too, Stu. @chrisjrn I believe that would mean taking my original suggestion at the top of the thread

Reply:
Great, thanks for all the feedback.